src/uwtools/tests/drivers/test_forecast.py

```diff
@@ -252,78 +271,115 @@ def test_stage_files(tmp_path, section, link_files):
     _, config_file, config = fv3_run_assets
+    expected_command = " ".join(fv3_mpi_assets)
     with patch.object(FV3Forecast, "_validate", return_value=True):
         with patch.object(forecast, "execute") as execute:
             fcstobj = FV3Forecast(config_file=config_file)
             with patch.object(fcstobj, "_config", config):
-                fcstobj.run()
-            execute.assert_called_once_with(cmd="srun --export=None test_exec.py")
+                fcstobj.run(cycle=dt.datetime.now())
+            execute.assert_called_once_with(cmd=expected_command)
 
 
-def test_FV3Forecast_run_dry_run(caplog, fv3_run_assets):
+def test_FV3Forecast__config_deleter():
+    config_file = fixture_path("forecast.yaml")
+    with patch.object(Driver, "_validate", return_value=True):
+        forecast = FV3Forecast(config_file=config_file)
+
+    assert forecast._config
+    del forecast._config
+    assert not forecast._config
+
+
+def test_FV3Forecast__config_setter():
+    config_file = fixture_path("forecast.yaml")
+    with patch.object(Driver, "_validate", return_value=True):
+        forecast = FV3Forecast(config_file=config_file)
+    new_config = {"foo": 1, "bar": 2}
+    forecast._config = new_config
+    assert forecast._config_data == new_config
+    assert forecast._config == new_config
+
+
+@pytest.mark.parametrize("with_batch_script", [True, False])
+def test_FV3Forecast_run_dry_run(capsys, fv3_mpi_assets, fv3_run_assets, with_batch_script):
     logging.getLogger().setLevel(logging.INFO)
     batch_script, config_file, config = fv3_run_assets
-    run_expected = """
-#!/bin/bash
-#SBATCH --account=user_account
-#SBATCH --nodes=1
-#SBATCH --ntasks-per-node=1
-#SBATCH --qos=batch
-#SBATCH --time=00:01:00
-srun --export=None test_exec.py
-""".strip()
+    if with_batch_script:
+        batch_components = [
+            "#!/bin/bash",
+            "#SBATCH --account=user_account",
+            "#SBATCH --nodes=1",
+            "#SBATCH --ntasks-per-node=1",
+            "#SBATCH --qos=batch",
+            "#SBATCH --time=00:01:00",
+        ] + fv3_mpi_assets
+        run_expected = "\n".join(batch_components)
+    else:
+        batch_script = None
+        run_expected = " ".join(fv3_mpi_assets)
+
     with patch.object(FV3Forecast, "_validate", return_value=True):
         fcstobj = FV3Forecast(config_file=config_file, dry_run=True, batch_script=batch_script)
         with patch.object(fcstobj, "_config", config):
-            fcstobj.run()
-    assert run_expected in caplog.text
+            print(fcstobj._config)
+            fcstobj.run(cycle=dt.datetime.now())
+    assert run_expected in capsys.readouterr().out
 
 
 def test_run_submit(fv3_run_assets):
     batch_script, config_file, config = fv3_run_assets
     with patch.object(FV3Forecast, "_validate", return_value=True):
-        with patch.object(forecast, "execute") as execute:
+        with patch.object(forecast, "submit_job") as submit_job:
```

Comment:
@maddenp-noaa I implemented this differently than you did in PR #302. I am struggling to figure out how to patch `self.scheduler.submit_job(outpath)` on line 175 of `drivers/forecast.py`. Any pointers?

Reply:
I think I got it. :)

Reply:
After instantiating `fcstobj`, try
```
with patch.object(fcstobj, "scheduler") as scheduler:
    ...
```
When the expression `self.scheduler.submit_job(outpath)` is then evaluated by the code-under-test:

1. `self.scheduler` should be a `Mock` object that will magically create any attributes/methods referenced on it, and record the references, so that
2. `self.scheduler.submit_job` will be _another_ `Mock` that will
3. record record that it was called with `outpath` as its argument.

After `fcstobj.run()` is called from your test, you should be able to e.g.
```
scheduler.submit_job.assert_called_once_with(output)
```
If you'd like for the magically-created `submit_job` `Mock` to return something specific (because the code-under-test does a `return self.scheduler.submit_job(outpath)`), you could
```
with patch.object(fcstobj, "scheduler") as scheduler:
    scheduler.submit_job.return_value = <whatever-you-want>    
```
In case it's not obvious, the reason we have to patch `_validate` on the `FV3Forecast` _class_ instead of on the `fcstobj` _object_ is because `_validate` is called from the constructor: We can't patch the instance object because we don't have one yet. In general, I think it's preferable to patch instances instead of classes, and we can do that in this case.

(BTW, there's special syntax in `unittest.mock` for patching multiple attributes on the same object. It's got some verbosity downsides itself, so I usually only use it if I have to patch at least 3 attributes on an object. If you get to like 3+ nested `with patch.object(...)` calls on the same object, it can pay to use it. Let me know if you ever want a demo on that -- or read the docs on `patch.multiple()`. I usually `from unittest.mock import DEFAULT as D` to make the `patch.multiple()` call a bit more concise. There might be some examples of this in `uwtools` already.)